Design System Reskin

KludgeKML opened the pull request:
Experimental visual refresh which updates Local Links Manager to use govuk_publishing_components elements and  Design System styles and components instead of the old Bootstrap-based ones. 

https://trello.com/c/JYROECMz/1592-reskin-local-links-manager

## Screenshots:

### Before:
<img width="1173" alt="Screenshot 2023-10-30 at 15 29 01" src="https://github.com/alphagov/local-links-manager/assets/4225737/707e3f7c-6522-4451-8f11-68c3bc08f1c2">

### After:
<img width="1194" alt="Screenshot 2024-02-26 at 16 03 13" src="https://github.com/alphagov/local-links-manager/assets/4225737/b9a04590-cc15-4e5f-9b8e-75ef76b57d8e">

### Before:
<img width="1157" alt="Screenshot 2023-10-30 at 15 29 28" src="https://github.com/alphagov/local-links-manager/assets/4225737/84a32fc8-4922-41e1-af30-fdbc7f4a4f81">

### After:
<img width="1186" alt="Screenshot 2024-02-26 at 16 45 05" src="https://github.com/alphagov/local-links-manager/assets/4225737/2a7983f5-a5a6-465c-a081-f515e59d321f">

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

KludgeKML committed:
Remove admin template gem, add publishing components gem

KludgeKML committed:
Update routes to replace admin template with publishing components

KludgeKML committed:
Remove obsolete initialiser

KludgeKML committed:
Add components initialiser

KludgeKML committed:
Remove unused stylesheets

KludgeKML committed:
Extract download/upload forms into separate pages

- Remove unnecessary JS
- Update Routes for new forms.
- Add component styles required
- Add design system styles
- Update controller to pass new parameters
  directly to exporter, and exporter to
  handle that.
- Remove public_send in exporter now that
  we're calling it with (sanitised) params.

KludgeKML committed:
Split out download links tests

KludgeKML committed:
Update exporter tests to handle new parameters

KludgeKML committed:
Update layout

- Update base layout
- Update JS and CSS with required components, remove obsolete JS/CSS
- Add application.css to manifest
- This is a publishing app, so use their more generous page width

KludgeKML committed:
Remove obsolete tabs partial

KludgeKML committed:
Update links controller and views

- Add shared general-purpose links table partial
- Add style for the edit links page
- Add colour tags for link status
- Update url presenter for use in the shared table
- Change link status alerts to use Warning Text
- Remove obsolete shared file
- Add visually hidden info to edit link in links table

Fix caps on "showing top" (SQUASH INTO LINKS)

KludgeKML committed:
Update Local Authorities controller, views, and model

- Remove obsolete shared local authorities partial
- split Homepage URL edit field into separate view/route
- Add sidebar styles for council pages
- Add metadata presenter code to LA presenter
- Add active scope to local_authority model
- Accept retired tag in index, default to only showing enabled

KludgeKML committed:
Update Services controller / views

- Add govuk page title/link to services index

KludgeKML committed:
Remove old favicons

KludgeKML committed:
Update links index tests

- Remove LGSL code expectation
- Remove links expectation
- Update broken link counter position/display
- Remove nav tabs expectation (put this somewhere more general)
- rename councils to avoid heisenbug where bbb appears first because it's in the assets digest filename.

KludgeKML committed:
Update tests for Service show/index

KludgeKML committed:
Update tests for LocalAuthority show/index

jon-kirwan commented on app/assets/stylesheets/components/app-side.scss:
```diff
@@ -0,0 +1,5 @@
+    @include govuk-responsive-margin(6, "top");
+    padding: govuk-spacing(3);
+    background: govuk-colour("light-grey");
+}
```
The indentation looks slightly incorrect here and I see that `stylelint` isn't logging any errors. I think `package.json` might need to be updated like so:
```
  "stylelint": {
    "extends": ["stylelint-config-gds/scss", "stylelint-stylistic/config"]
  },
  "devDependencies": {
    "standardx": "^7.0.0",
    "stylelint": "^15.11.0",
    "stylelint-config-gds": "^1.1.1",
    "stylelint-stylistic": "^0.4.5"
  }
```
which should then log a few indentation Sass errors.

KludgeKML replied:
Ah - thanks! Have fixed that, and added the lint update as a final commit.

jon-kirwan replied:
Excellent thanks!

As an aside, I wonder if we should include some view formatting. I notice that [Whitehall](https://github.com/alphagov/whitehall/blob/main/Gemfile#L62) use one https://github.com/Shopify/erb-lint which does all the usual linter type things - removing empty lines, standardising quotations etc.

It's fairly easy to setup. You can copy the yml file https://raw.githubusercontent.com/alphagov/whitehall/main/.erb-lint.yml.

Then run like so: ` bundle exec erblint * -a`.

KludgeKML replied:
Ah, nice. Have added that to the CI in the last commit (and retroactively applied its fix to the various previous commits)

jon-kirwan replied:
Great thanks!

jon-kirwan commented on app/views/shared/_links_table.html.erb:
```diff
@@ -0,0 +1,54 @@
+      { text: "<span class=\"govuk-tag govuk-tag--#{pres.status_tag_colour}\">#{pres.status_description}</tag>".html_safe },
+      { text: pres.last_checked },
+      { text: link_to("Edit", edit_link_path(link.local_authority, link.service, link.interaction)) },
+    ]
```
A couple of places here where we can add the Design System link formatting `govuk-link`:
```
{ text: link_to(link.local_authority.name, local_authority_path(link.local_authority), class: 'govuk-link') },
...
{ text: link_to("Edit", edit_link_path(link.local_authority, link.service, link.interaction), class: 'govuk-link') },
```

KludgeKML replied:
Done!

jon-kirwan commented on app/views/services/index.html.erb:
```diff
@@ -2,45 +2,35 @@
+  @table_rows = @services.map do |service|
+    [
+      { text: link_to(service.label, service_path(service, filter: 'broken_links')) },
+      { text: service.lgsl_code },
```
We can add the Design System link formatting `govuk-link` here:
```
{ text: link_to(service.label, service_path(service, filter: 'broken_links', class: 'govuk-link')) },
```

KludgeKML replied:
Done!

jon-kirwan commented on app/views/local_authorities/index.html.erb:
```diff
@@ -1,40 +1,36 @@
+  @table_rows = @authorities.map do |authority|
+    [
+      { text: link_to(authority.name, local_authority_path(authority.slug, filter: 'broken_links')) },
+      { text: authority.active? ? "Yes" : "No" },
```
We can add the Design System link formatting `govuk-link` here:
```
{ text: link_to(authority.name, local_authority_path(authority.slug, filter: 'broken_links', class: 'govuk-link')) },
```

KludgeKML replied:
Done!

jon-kirwan commented on app/views/services/index.html.erb:
```diff
@@ -2,45 +2,35 @@
+  text: "Services (#{@services.count})",
+  margin_bottom: 5,
+} %>
 
```
Now that 'Local Links Manager' has been included into the header, it appears that all page headings are formatted as `h2`. We should probably change them all to `h1`.
```
<%= render "govuk_publishing_components/components/heading", {
  text: "Services (#{@services.count})",
  heading_level: 1,
  font_size: 'l',
  margin_bottom: 5,
} %>
```

KludgeKML replied:
Done!

KludgeKML committed:
Improve linting

jon-kirwan commented on app/views/services/index.html.erb:
```diff
@@ -2,45 +2,35 @@
+    },
+    {
+      text: "LGSL Code",
+    },
```
Is "LGSL" widely understood? Any need to write it out in full with the abbreviation in brackets?

KludgeKML replied:
Hmm, probably not. I'll write it out.

KludgeKML replied:
Oh, I think maybe we shouldn't write it out for this table, it'll make the column enormous. But I'll expand on it in the show template.

jon-kirwan commented on app/views/links/edit.html.erb:
```diff
@@ -1,44 +1,49 @@
+        destructive: true,
+      }
+    %>
+    <% end %>
```
The adjacent colours (of the surrounding block and the 'delete' button) might be difficult to distinguish (https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast).

We could:
- change the colour of `app-danger-block` to something else (like `govuk-colour("light-grey")`) or
- remove the background colour altogether. The Design System seems to only include examples of using destructive button types on a white background https://design-system.service.gov.uk/components/button/warning/.

jon-kirwan commented on app/views/links/edit.html.erb:
```diff
@@ -1,44 +1,49 @@
 
 <%= render partial: "link_status_alerts", locals: { alerts: @link.link_warnings, html_class: "warning", prefix: "Warning" } %>
 <%= render partial: "link_status_alerts", locals: { alerts: @link.link_errors, html_class: "danger", prefix: "Error" } %>
 
```
I think the alerts partial uses the Bootstrap alert component (`alert alert-`). It looks like the notification banner is the intended replacement for alerts https://design-system.service.gov.uk/components/notification-banner/ and https://components.publishing.service.gov.uk/component-guide/notice.

KludgeKML committed:
Add ERB linter

- as per Whitehall, if run in dev environment, autocorrect by default
- Add ERB linting step in CI

jon-kirwan commented on app/views/shared/_link_table_row.html.erb:
```diff
@@ -1,10 +1,9 @@
+      link.updated? ? "success" : "",
+    ].join(" "),
+    data: link.row_data do %>
   <td class="text-center lgsl">
```
I'm pretty sure the class names `first-row` and `success` are old formatting styles and we can safely delete them.

KludgeKML replied:
Done!

jon-kirwan commented on app/assets/stylesheets/application.scss:
```diff
@@ -1,42 +1,19 @@
+@import "govuk_publishing_components/components/breadcrumbs";
+@import "govuk_publishing_components/components/button";
+@import "govuk_publishing_components/components/checkboxes";
+@import "govuk_publishing_components/components/heading";
```
We'll need to include the form file upload component styles here: `govuk_publishing_components/components/file-upload` to see the input styles on this page https://local-links-manager.integration.publishing.service.gov.uk/local_authorities/north-tyneside/upload_links_form

KludgeKML replied:
Done

jon-kirwan commented on app/views/local_authorities/show.html.erb:
```diff
@@ -1,39 +1,31 @@
+              link_to("Download Links", download_links_form_local_authority_path(@authority)),
+              link_to("Upload Links", upload_links_form_local_authority_path(@authority)),
+            ],
+          } %>
```
Probably just need to add the `govuk-link` formatting here. You could also add `margin_bottom: 0` to the list component to have equal vertical spacing.

KludgeKML replied:
Done

jon-kirwan commented on app/views/links/_link_table_row.html.erb:
```diff
@@ -1,7 +1,6 @@
-    data: link.row_data do
-%>
+<%= content_tag "tr",
+    class: link.updated? ? "success" : "",
```
Some more unneeded styles here. We can delete `class: link.updated? ? "success" : "",`

KludgeKML replied:
Ah, I've actually got rid of this entire file, it's not necessary now I just haven't squashed the deletion. Let me just tidy up the tinkering commits.

jon-kirwan commented on app/assets/stylesheets/components/app-side.scss:
```diff
@@ -0,0 +1,5 @@
+.app-side {
+  @include govuk-responsive-margin(6, "top");
```
We can remove `@include govuk-responsive-margin(6, "top");` to align with the metadata block.

KludgeKML replied:
Done.

jon-kirwan commented on app/views/shared/_links_table.html.erb:
```diff
@@ -0,0 +1,55 @@
+    },
+    {
+      text: "Last Checked",
+    },
```
I'm wondering if we can remove the 'last checked' column on the table page. The information mostly seems to be 'one year ago' or 'not checked'. The same information is available on the 'show' page. Removing it would allow for more space in the table.

KludgeKML replied:
Hmm. If you look at it on production or integration it's more like "8 hours ago" etc. I'll agree it's probably not that useful though. We can always put it back if necessary.

KludgeKML replied:
Done!

KludgeKML committed:
Update shared URL Status Presensation tests

KludgeKML committed:
fix security bug in exporter subclasses

1pretz1 commented on app/views/shared/_links_table.html.erb:
```diff
@@ -0,0 +1,51 @@
+<%
```
It feels a bit strange to have a partial that is mostly Ruby logic (line 1 to 44). I think moving this to a defined presenter class or helper would be a bit more idiomatic.

KludgeKML replied:
Moved code into a presenter

1pretz1 commented on app/views/local_authorities/index.html.erb:
```diff
@@ -1,40 +1,38 @@
 
-<%= render partial: 'shared/home_nav' %>
+<%
+  @table_rows = @authorities.map do |authority|
```
Similar comment as the last, it feels like it would be more natural for this logic to live outside the view.

KludgeKML replied:
Moved code into a presenter

1pretz1 commented on app/views/services/index.html.erb:
```diff
@@ -1,46 +1,46 @@
 
-<%= render partial: 'shared/home_nav' %>
+<%
+  @table_rows = @services.map do |service|
```
And here

KludgeKML replied:
Moved code into a presenter.

1pretz1 commented on app/assets/javascripts/download_checkboxes.js:
```diff
@@ -1,17 +0,0 @@
-document.addEventListener('DOMContentLoaded', function (event) {
```
Nice!

1pretz1 reviewed (commented):
Changes look great Keith, fantastic work! I've just left one comment on the code itself which is concerned with moving pure Ruby logic out of the view.

I've then deployed the changes to integration and noted a few other small things visually:

1) We've removed the feature to view all links or just broken links for a service / authority, do we intend to add the filter back? No biggie if not. 

2) A related point is that we've added 'visits this week' to service / LA which is great, could we set the default ordering to visits, like we do on the main broken links page?   Feels like the most visits should be at the top to give the potentially broken link the best visibility. 

3) broken links title: Broken Links (Showing top 200 of 20931) -> Broken Links (showing top 200 of 20931)

4) The metadata under a council / service is a bit hard to distinguish between the key / value. Maybe we can use the [summary list component](https://deploy-preview-3266--govuk-design-system-preview.netlify.app/components/summary-list/) for these bits of information?
<img width="1246" alt="Screenshot 2024-02-23 at 17 16 43" src="https://github.com/alphagov/local-links-manager/assets/24479188/a005af38-67d6-4a94-9151-e5f9b553bc4a">
<img width="1278" alt="Screenshot 2024-02-23 at 17 16 54" src="https://github.com/alphagov/local-links-manager/assets/24479188/e8dbc079-59a5-4a22-8e92-392f8ca6772b">

KludgeKML commented:
@1pretz1 
1. It's been removed, so there's a temporary loss of functionality on those two lists, but there's a larger question about filtering that I'm planning to address when this is merged. If we think the temporary loss is too annoying, we could add in a simple link to toggle broken/all?

2. It's slightly more awkward to sort by traffic on the other two lists, so I thought it would be reasonable to leave the sort order as it currently is. I'm experimenting with giving the tables sortable headers.

3. Done

4. Possibly better - I'm going to make an experimental branch off this one to check.

1pretz1 reviewed (commented):
Thanks for making those changes Keith! Table presenter is a good addition and the new summary list looks great! Approving as nothing major to block anymore, I do have a few suggestions below which can either be rolled into another PR (as this one is a bit large). 

I had one more minor point on the interaction table column, I think it looks a bit strange as the font / line height is smaller and greyed. Making the font size consistent would be good and possibly the colour (as it is v similar to the black full size).

> It's been removed, so there's a temporary loss of functionality on those two lists, but there's a larger question about filtering that I'm planning to address when this is merged. If we think the temporary loss is too annoying, we could add in a simple link to toggle broken/all?

Playing about with the UI, I'd probably say not being able to filter by broken links is a bit annoying. A toggle sounds a good idea in the meantime, or if the filtering work is on the way soon then I think it's probably okay as is for now. 

> It's slightly more awkward to sort by traffic on the other two lists, so I thought it would be reasonable to leave the sort order as it currently is. I'm experimenting with giving the tables sortable headers.

I think ordering by views is definitely the most useful sort option but not a blocker if not poss + sortable table headers sounds good!

KludgeKML committed:
Update LinkFilterHelper with filter toggle util method

jon-kirwan reviewed (approved):
Great work, well done!
